test/test_controller.py

```diff
@@ -1282,3 +1283,198 @@ def test_collect_binlogs_to_purge():
+
+        m_controller.mark_backup_requested(backup_reason=BackupStream.BackupReason.requested)
+        data_generator.stop()
+
+        def all_backups_completed(expected_num):
+            backups = m_controller.state["backups"]
+            assert len(backups) == expected_num
+            assert all(backup["completed_at"] is not None for backup in backups)
+
+        while_asserts(lambda: all_backups_completed(expected_num=2), timeout=30)
+
+        m_controller.stop()
+
+        # get the most recent backup (corrupted)
+        corrupted_backup = sort_completed_backups(m_controller.state["backups"])[1]
+
+        new_controller = build_controller(
+            Controller,
+            default_backup_site=default_backup_site,
+            mysql_config=mysql_empty,
+            session_tmpdir=session_tmpdir,
+        )
+
+        new_controller.start()
+
+        restore_operation_side_effects = LimitedSideEffectMock(
+            original_function_to_mock=BasebackupRestoreOperation.restore_backup,
+            max_calls_with_side_effects=2,
+        )
+
+        # raise exceptions for all attempts to restoring corrupted basebackup,
+        # when trying to restore the previous basebackup, it should work as we expect
+        with patch.object(
+            BasebackupRestoreOperation,
+            "restore_backup",
+            autospec=True,
+            side_effect=restore_operation_side_effects,
+        ):
+            wait_for_condition(lambda: new_controller.state["backups_fetched_at"] != 0, timeout=2)
+
+            new_controller.restore_backup(site=corrupted_backup["site"], stream_id=corrupted_backup["stream_id"])
+
+            def restoration_has_phase(phase):
+                return new_controller.restore_coordinator and new_controller.restore_coordinator.state["phase"] == phase
+
+            # second basebackup should fail
+            wait_for_condition(lambda: restoration_has_phase(RestoreCoordinator.Phase.failed_basebackup), timeout=10)
+
+            # after second basebackup fails, it should try to restore the first backup we generated
+            # this one should be successfull
+            wait_for_condition(lambda: restoration_has_phase(RestoreCoordinator.Phase.completed), timeout=30)
+
+            new_controller._refresh_backups_list()  # pylint: disable=protected-access
+
+            current_backups = sort_completed_backups(new_controller.state["backups"])
+            corrupted_backup = current_backups[1]
+            # failed basebackup should be marked as broken
+            assert corrupted_backup["broken_at"] is not None
+
+            # verify the empty mysql restored all data.
+            with mysql_cursor(**mysql_empty.connect_options) as cursor:
+                expected_row_count = data_generator.row_count
+                cursor.execute("SELECT COUNT(*) AS count FROM db1.t1")
+                result_row_count = cursor.fetchone()["count"]
+                assert expected_row_count == result_row_count
```

Comment:
this is another assertion, different from "broken backup is marked as such". 

how does this related to the other assertion? there're multiple backups, some are broken, some are not? We should make it clear what we're testing.

If the backup is broken, we should NOT be able to query the rows, right? OTC, if some restoration succeeded, then we should get the data. 

Maybe (that's just what I'm hypothesizing) you wanted to verify that, if the broken backup holds different data than the non-broken one, the non-broken one is the one that actually gets restored. But I don't see this kind of assertion, if I understand the code properly.

Also: the monkeypatching is far from ideal; the best thing would being able to make multiple backups, then corrupt one of them so that our code actally throws an exception.

Another path to explore: split tests. In one test, you just assert that, if a backup is marked as broken, its content is not restored (you can try marking it as broken "manually" somehow), by verifying what's the state of the db after the test.

In another test, you verify that, if an exception happens during restore, then such backup is marked as broken, there, you don't verify the final db content because you don't care.

Reply:
so it goes like this:

1. Attempt to restore a "corrupted" backup
2. Fail all attempts  (first 2 side effects)
3. RestoreController enters into "failed_basebackup" phase
4. Controller tries to restore previous backup
5. RestoreController retries with that backup (we assume this one is healthy) (no side effects here)
6. RestoreController successfully restores backup and enters into "completed" phase

So, that's why the 2 `wait_for_condition`: (1st one checks attempt with corrupted backup and 2nd with healthy backup)
https://github.com/aiven/myhoard/blob/eb829289b3292963ac1b7211a75264ddc04c0e77/test/test_controller.py#L1360-L1365


And yes, I did tried actually corrupting the xbstream file from the basebackup, but it was quite ugly. And IMHO this isn't the test that should cover it, rather do a separate test under `test_restore_coordinator`.